File: Engine/Leviathan/GameObjects/Component.cs
namespace Leviathan.GameObjects
{
	public class Component
	{
		protected GameObject? GameObject { get; set; }

		internal void SetGameObject(GameObject _gameObject) => GameObject = _gameObject;

		public virtual void Start(params object[] _data) { }

		public virtual void Tick() { }

		public virtual void PhysicsTick(float _timeStep) { }

		public virtual void Render() { }

		public virtual void OnDestroy() { }

		public virtual void OnRenderGizmos() { }
	}
}

File: Dependencies/include/Leviathan/GameObjects/Component.cpp
#include "Component.h"

#include <Leviathan/GameObjects/GameObject.h>

namespace Leviathan
{
	TObjectPtr<GameObject> Component::GetOwner()
	{
		TObjectPtr<GameObject> ptr = TObjectPtr<GameObject>(m_owner);
		ptr.SetCopied(true);

		return TObjectPtr<GameObject>(ptr);
	}
}
